### PR TITLE
frontend: Strengthen Python code check

### PR DESCRIPTION
**Pull request by georgemitenkov**
TL;DR:
We allow only two ways of creating a front-end program:
```python
def foo():
  # no blocks, ops only!

  # explicit return
  return
```
and
```python
def foo():
  @block
  def bb0():
    # ops only!

    # explicit return
    return

  # explicit entry block call
  return bb0()
```
The only exception are compile-time constants which are inlined into functions/blocks and therefore do not affect code generation.

Previously, the code check was not great: 
1) Code duplication
2) Not too restrictive and wrong wrt blocks (e.g. blocks did not carry explicit terminators)
3) Similar to 2), functions were not required to have a terminator
4) We were allowing grouping of ops in functions, and without them

Now, we only allow functions in the program. The reason is: analysis is simpler, if you want operations then you can just put them inside a dummy function. Also, both functions and blocks need terminators. This way we can make more assumptions for code generation, e.g. that the last statement in the function is a `return`, etc. If we ever want to add more fancy things (like not specifying void return), this can be implemented by adapting the AST but keeping code generation logic intact.

This also fixes basic blocks. Now each block has to return either a call to another block, and if expression with calls to other blocks, or otherwise it is considered as a terminating block for a function. In case 0 or >1 terminating blocks exist, we abort. Again, this make code generation for blocks (not supported yet) way easier and nicer.

Lastly, there are no real changes to code generation pipeline, only the checks before. This will be fixed in a next patch so that it is easier to review.

### CODE REVIEW

File: xdsl/frontend/python_code_check.py [hunk 1]
```diff
@@ -1,176 +1,329 @@
...
-    function_and_block_names: Dict[str, Set[str]] = field(default_factory=dict)
-    """Tracks duplicate function names and duplicate block labels."""
+            # Otherwise we allow anything, and only have to carefully look at
+            # inner functions.
+            if not isinstance(stmt, ast.FunctionDef):
+                continue
 
-    def visit(self, node: ast.AST) -> None:
-        super().visit(node)
+            # Only blocks are alloed
```
Comment (webmiche) [inline xdsl/frontend/python_code_check.py:290]:
```suggestion
            # Only blocks are allowed
```

**Review by webmiche — APPROVED** [review]
LGTM. I feel it is a good idea to focus on the basics first and then build the rest on top of it. We could also jsut transform the AST to add the return statements before calling this functionality (in a later step).

**Comment by georgemitenkov** [comment]
@webmiche precisely! 

I also though so too, e.g. we can also make any function use explicit entry block by transforming the AST. But then the error messages become tricky because these new statements are not associated with any source line.

If we can overcome issue with errors, we should definitely do the AST wrapping approach because then all programs will be uniform :)

**Comment by webmiche** [comment]
Indeed, that seems to be an issue. Feels like we would have to catch exceptions and change their linenumbers somehow. I guess you could also recommend adding the proper returns and stuff as aprt of the exception print, s.t. people get proper errors once they reexecute...

File: xdsl/frontend/python_code_check.py [hunk 2]
```diff
@@ -189,18 +342,15 @@ class CheckAndInlineConstants:
...
     # here b = 25
     ```
-
-    Note that the algorithm does not remove constant definitions from the AST,
-    but this functionality can be added later.
     """
 
     @staticmethod
-    def run(stmts: Sequence[ast.stmt], file: str | None) -> None:
+    def run(stmts: list[ast.stmt], file: str | None) -> None:
```
Comment (superlopuh) [inline xdsl/frontend/python_code_check.py:348]: why this change?

Comment (georgemitenkov) [reply xdsl/frontend/python_code_check.py:348]: Should I keep it as a sequence? I was working on the old branch with `List[ast.stmt]` and as I can remember we wanted to have `list` instead?

Comment (superlopuh) [reply xdsl/frontend/python_code_check.py:348]: `list` is better than `List`, but Sequence is generally easier to deal with, primarily because it's covariant so accepts lists of subclasses

**Review by superlopuh — APPROVED** [review]

File: xdsl/frontend/python_code_check.py (hunk 2 above)
Comment (georgemitenkov) [reply xdsl/frontend/python_code_check.py:348]: Let me use a Sequence then instead.
